### PR TITLE
Support required property.

### PR DESCRIPTION
**Pull request by pengweiqhca**

### CODE REVIEW

File: src/Autofac.Integration.Web/Forms/AttributedInjection.cs [hunk 1]
```diff
@@ -10,6 +12,9 @@ namespace Autofac.Integration.Web.Forms;
 /// </summary>
 internal class AttributedInjection : PageInjectionBehavior
 {
+    private static readonly ConcurrentDictionary<Type, bool> HasRequiredMemberAttributeCache =
```
Comment (alistairjevans) [inline src/Autofac.Integration.Web/Forms/AttributedInjection.cs]: Could you see if you can use the Autofac shared reflection cache? It allows caches to be invalidated per-assembly if required.

For example:

```
HasRequiredMemberAttributeCache = ReflectionCacheSet.Shared.GetOrCreateCache<ReflectionCacheDictionary<Type, bool>>("hasrequiredmemberattributecache");
```

This still gives you basically the same functionality, but we can centrally control it as needed.

File: src/Autofac.Integration.Web/Forms/AttributedInjection.cs [hunk 2]
```diff
@@ -33,10 +38,30 @@ protected override Func<object, object> GetInjector(IComponentContext context)
...
             else
             {
                 return target;
             }
         };
     }
+
+    private static bool HasRequiredMemberAttribute(Type type)
+    {
+        return HasRequiredMemberAttributeCache.GetOrAdd(type, t =>
```
Comment (alistairjevans) [inline src/Autofac.Integration.Web/Forms/AttributedInjection.cs:61]: Can you please bring over the comment from Autofac to explain why we have to walk the tree?

```
// The RequiredMemberAttribute (may)* have Inherit = false on its AttributeUsage options,
// so walk the tree.
// (*): see `HasRequiredMemberAttribute` doc for why we dont really know much about the concrete attribute.
```

File: test/Autofac.Integration.Web.Test/Forms/AttributedInjectionModuleFixture.cs [hunk 3]
```diff
@@ -95,6 +100,21 @@ public void PropertyNotInjectedWhenValueSet()
         Assert.Equal(ExplicitlyProvidedString, page.Property);
     }
 
+        [Fact]
```
Comment (alistairjevans) [inline test/Autofac.Integration.Web.Test/Forms/AttributedInjectionModuleFixture.cs]: Can you please fix the indentation here?

File: test/Autofac.Integration.Web.Test/Forms/AttributedInjectionModuleFixture.cs [hunk 4]
```diff
@@ -36,6 +36,11 @@ private class UnsetPropertyInjectedPage : HttpHandler
         public string Property { get; set; }
     }
 
+        private class RequiredMemberInjectedPage : HttpHandler
```
Comment (alistairjevans) [inline test/Autofac.Integration.Web.Test/Forms/AttributedInjectionModuleFixture.cs]: Indentation here too please.

File: src/Autofac.Integration.Web/ContainerProviderContainer.cs [hunk 5]
```diff
@@ -102,6 +102,17 @@ public ILifetimeScope BeginLifetimeScope(object tag, Action<ContainerBuilder> co
...
+    /// <summary>
+    /// Resolve an instance of the provided registration within the context.
+    /// </summary>
+    /// <param name="request">The resolve request.</param>
+    /// <returns>
+    /// The component instance.
+    /// </returns>
+    /// <exception cref="ComponentNotRegisteredException"/>
+    /// <exception cref="DependencyResolutionException"/>
+    public object ResolveComponent(ResolveRequest request) => _containerProvider.RequestLifetime.ResolveComponent(request);
```
Comment (alistairjevans) [inline src/Autofac.Integration.Web/ContainerProviderContainer.cs]: Why was this added?

**Comment by alistairjevans** [comment]
@pengweiqhca, I'm not sure I understand why the existing support for required property injection in Autofac 7+ doesn't cover this case? https://autofac.readthedocs.io/en/latest/register/prop-method-injection.html#required-properties

Can you explain?

**Comment by pengweiqhca** [comment]
ASP.NET Web Forms only supports [property injection](https://github.com/autofac/Autofac.Web/blob/develop/src/Autofac.Integration.Web/Forms/DependencyInjectionModule.cs#L48).

**Comment by alistairjevans** [comment]
Ah, of course, good point! Been a while since I've looked at the WebForms stuff.

**Review by alistairjevans — CHANGES_REQUESTED** [review]
Just a couple of points, thanks!

File: src/Autofac.Integration.Web/ContainerProviderContainer.cs (hunk 5 above)
Comment (pengweiqhca) [reply src/Autofac.Integration.Web/ContainerProviderContainer.cs]: Sorry, when I was modifying the code, I could not compile it and forgot to delete it after pulling the latest code.

**Comment by pengweiqhca** [comment]
All things are resolved.

**Review by alistairjevans — APPROVED** [review]
Looks good, thanks, sorry for the delay.

**Pull request merged**